### PR TITLE
MNIST tutorial improvements.

### PR DESCRIPTION
**Pull request by madjam**

### CODE REVIEW

File: docs/tutorials/python/mnist.md [hunk 1]
```diff
@@ -1,116 +1,135 @@
 # Handwritten Digit Recognition
 
-This tutorial guides you through a classic computer vision application: identify
-hand written digits with neural networks.
+In this tutorial, we’ll give you a step by step walk-through of how to build a hand written digit classifier using the [MNIST](https://en.wikipedia.org/wiki/MNIST_database) dataset. For someone new to deep learning, this exercise is arguably the “Hello World” equivalent.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "hand written -> "hand-written" (two-word adjective)

File: docs/tutorials/python/mnist.md [hunk 2]
```diff
@@ -1,116 +1,135 @@
...
 
-## Loading Data
+MNIST is a widely used dataset for the hand written digit classification task. It consists of 70,000 labeled 28×28 pixel grayscale images of hand written digits. The dataset is split into 60,000 training images and 10,000 test images. There are 10 classes (one for each of the 10 digits). The task at hand is to train a model using the 60,000 training images and subsequently test its classification accuracy on the 10,000 test images.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "hand written" -> "hand-written"

File: docs/tutorials/python/mnist.md [hunk 3]
```diff
@@ -1,116 +1,135 @@
...
+
+Image batches are commonly represented by a 4-D matrix with shape `(batch_size, num_channels, width, height)`. For MNIST dataset, since the images are grayscale, there is only one color channel. Also, the images are 28x28 pixels, and so each image has width and height equal to 28. Therefore, the shape of input is `(batch_size, 1, 28, 28)`. Another important consideration is the order of input samples. When feeding training examples, it is critical that we don't feed samples with the same label in succession. Doing so can slow down training.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: change to "4-D array" (a matrix is by definition 2D)

Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "For MNIST dataset" -> "For the MNIST dataset"

File: docs/tutorials/python/mnist.md [hunk 4]
```diff
@@ -1,116 +1,135 @@
...
+Data iterators take care of this by randomly shuffling the inputs. Note that we only need to shuffle train data. The order does not matter for test data.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "only need to shuffle train data" -> "only need to shuffle the training data"

File: docs/tutorials/python/mnist.md [hunk 5]
```diff
@@ -1,116 +1,135 @@
...
+
+The following source code initializes the data iterators for MNIST dataset. Note that we initialize two iterators: one for train data and one for test data.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "for MNIST dataset" -> "for the MNIST dataset"

Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "one for the training data and one for test data"

File: docs/tutorials/python/mnist.md [hunk 6]
```diff
@@ -1,116 +1,135 @@
...
 
 ```python
 batch_size = 100
 train_iter = mx.io.NDArrayIter(mnist['train_data'], mnist['train_label'], batch_size, shuffle=True)
 val_iter = mx.io.NDArrayIter(mnist['test_data'], mnist['test_label'], batch_size)
 ```
 
+## Training
+We will cover a couple of approaches for performing the hand written digit recognition task. The first approach makes use of a more traditional deep neural network architecture called Multilayer Percepton. We'll discuss its drawbacks and use that as a motivation to introduce a second more advanced approach called Convolution Neural Network that has proven to work very well for image classification tasks.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: "more traditional" -> "traditional" (we can't compare it to anything because we haven't introduced another model yet)

Comment (zackchase) [inline docs/tutorials/python/mnist.md]: change to "called the Multilayer Perceptron (MLP)."

Comment (zackchase) [inline docs/tutorials/python/mnist.md]: -> "Convolutional Neural Network (CNN)"

File: docs/tutorials/python/mnist.md [hunk 7]
```diff
@@ -1,116 +1,135 @@
...
+
 ## Multilayer Perceptron
 
-We first use [multilayer perceptron](https://en.wikipedia.org/wiki/Multilayer_perceptron) to solve this problem. We
-define a multilayer perceptron by using MXNet's symbolic interface. The
-following command create a place holder variable for the input data.
+The first approach makes use of a [Multilayer Perceptron](https://en.wikipedia.org/wiki/Multilayer_perceptron), MLP for short, to solve this problem. We'll define the MLP using MXNet's symbolic interface. We begin by creating a place holder variable for the input data. When working with an MLP, we need to flatten our 28x28 images into a flat 1-D structure of 784 (28 * 28) raw pixel values. The order of pixel values in the flattened vector does not matter as long as we are being consistent about how we do this across all images.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: Redundant since we just said that above. Maybe remove the redundant part under training. 

Just say "Multilayer Perceptron (MLP)". this is the standard way to introduce an acronym. Other wise don't capitalize the term.

File: docs/tutorials/python/mnist.md [hunk 8]
```diff
@@ -1,116 +1,135 @@
...
 
 ```python
 data = mx.sym.var('data')
 # Flatten the data from 4-D shape into 2-D (batch_size, num_channel*width*height)
 data = mx.sym.flatten(data=data)
 ```
+One might wonder if we are discarding valuable information by flattening. That is indeed true and we'll cover this more when we talk about convolutional neural networks where we preserve the input shape. For now, we'll go ahead and work with flattened images.
+
+MLP contains several fully connected layers. A fully connected layer or FC layer for short, is one where each neuron in the layer is connected to every neuron in its preceding layer. From a linear algebra perspective, an FC layer applies an [affine transform](https://en.wikipedia.org/wiki/Affine_transformation) to the *n x m* input matrix *X* and outputs a matrix *Y* of size *n x k*, where *k* is the number of neurons in the FC layer. *k* is also referred to as the hidden size. The output *Y* is computed according to the equation *Y = W X + b*. The FC layer has two learnable parameters, the *m x k* weight matrix *W* and the *m x 1* bias vector *b*.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: -> "MLPs contain"

File: docs/tutorials/python/mnist.md [hunk 9]
```diff
@@ -1,116 +1,135 @@
...
+
 
-A multilayer perceptron contains several fully-connected layers. A fully-connected
-layer, with an *n x m* input matrix *X* outputs a matrix *Y* with size *n x k*,
-where *k* is often called as the hidden size. This layer has two learnable parameters, the
-*m x k* weight matrix *W* and the *m x 1* bias vector *b*. It compute the
-outputs with *Y = W X + b*.
+In an MLP, the output of most FC layers is fed into an activation function, which applies an element-wise non-linearity. This step is critical and it gives neural networks the ability to classify inputs that are not linearly separable. Common choices for activation functions are sigmoid, tanh, and [rectified linear unit](https://en.wikipedia.org/wiki/Rectifier_%28neural_networks%29) ("ReLU" for short). In this example, we'll use the ReLU activation function which has several desirable properties and is typically considered a default choice.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: -> "the outputs of most FC layers are fed intro"

Comment (zackchase) [inline docs/tutorials/python/mnist.md]: Again introduce the acronym "Rectified Linear Unit (ReLU)". No need to say "for short"

File: docs/tutorials/python/mnist.md [hunk 10]
```diff
@@ -1,116 +1,135 @@
...
 
-The output of a fully-connected layer is often feed into an activation layer,
-which performs element-wise operations. Common activation functions include
-sigmoid, tanh, and rectifier (or "relu").
+The following source code declares two fully connected layers with 128 and 64 neurons each. Furthermore, these FC layers are sandwiched between ReLU activation layers each one responsible for performing an element-wise ReLU transformation on the FC layer output.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: Maybe just say "code". source code sounds a bit too formal

File: docs/tutorials/python/mnist.md [hunk 11]
```diff
@@ -1,116 +1,135 @@
...
+acc = mx.metric.Accuracy()
+mlp_model.score(test_iter, acc)
+print(acc)
+assert acc.get()[1] > 0.96
+```
+If everything went well, we should see an accuracy value that is around 0.96, which means that we are able to accurately predict the digit in 96% of test images. This is a pretty good result. But as we will see in the next part of this tutorial, we can do a lot better than that.
+
+## Convolutional Neural Network
+
+Earlier we briefly touched on a drawback of MLP when we said we need to discard the input image's original shape and flatten it as a vector before we can feed it as input to the MLP's first fully connected layer. Turns out this is an important issue because we don't take advantage of the fact that pixels in the image have natural spatial correlation along the horizontal and vertical axes. A convolutional neural network (CNN) aims to address this problem by using a more structured weight representation. Instead of flattening the image and doing a simple matrix-matrix multiplication, it employs one or more convolutional layers that each performs a 2-D convolution on the input image.
```
Comment (zackchase) [inline docs/tutorials/python/mnist.md]: -> "Earlier," (missing comma)

**Comment by madjam** [comment]
@zackchase thanks for reviewing this. Pushed updates to address your comments.
